packages/doenetml-worker-rust/lib-doenetml-core/src/core/general_state_var/string_state_var.test.rs

```diff
@@ -0,0 +1,209 @@
+    ////////////////////////////////////////////////////////////////
+
+    // on the state variable view, record that we request the value be false
+    state_var_view.queue_update(String::from("Bye World"));
+
+    let invert_result = state_var.invert(false);
+
+    assert!(invert_result.is_err());
+
+    ////////////////////////////////////////////////////////////////
+    // Step 5: change the text children to make it "Bye Earth"
+    ////////////////////////////////////////////////////////////////
+
+    child1_var.set_value(String::from("Bye "));
+    child2_var.set_value(String::from("Earth"));
+
+    state_var.calculate_and_mark_fresh();
+
+    // now the value should be "Bye Earth"
+    assert_eq!(state_var.get().clone(), "Bye Earth");
+}
+
+/// Testing the case of string state variable based on an attribute
+/// For example, if we have a `<textInput prefill/>`
+/// its `prefill` state variable would follow this pattern.
+#[test]
+fn string_state_var_with_attribute() {
+    let mut state_var =
+        StringStateVar::new_from_attribute("my_attr", String::from("")).into_state_var();
+
+    //////////////////////////////////////////////////
+    // check that get the correct data queries
+    //////////////////////////////////////////////////
+    let queries = state_var.return_data_queries();
+
+    // should return a query for "my_attr" with text children
+    assert_eq!(
+        queries,
+        vec![DataQuery::AttributeChild {
+            attribute_name: "my_attr",
+            match_profiles: vec![ComponentProfile::Text],
+            always_return_value: true,
+        },]
+    );
+
+    // Note: from this point on, at the level of these tests where dependencies are specified manually,
+    // there is no difference in this state variable as the above ones with children,
+    // so we don't repeat the same steps
+}
+
+/// If the variable is based on a single variable using its default value,
+/// the variable should also be designated as coming from default.
+/// For example, if we have a `<text></text>`,
+/// its `value` state variable should be the default
+#[test]
+fn propagate_came_from_default() {
+    // create a string state variable requesting children
+    let mut state_var = StringStateVar::new_from_children(String::from("")).into_state_var();
+
+    // need to return data queries since side effect is saving the required data
+    state_var.return_data_queries();
+
+    let (child_dependency, _child_var) = create_state_var_dependency(String::from(""), true);
+
+    let dependencies_created_for_data_queries =
+        vec![DependenciesCreatedForDataQuery(vec![child_dependency])];
+
+    state_var.save_dependencies(&dependencies_created_for_data_queries);
+
+    state_var.calculate_and_mark_fresh();
+
+    // we expect a value of "hello" and came_from_default to be false,
+    // since that the variable is coming from the child dependency, which we initialized to "hello".
+    assert_eq!(state_var.get().clone(), "");
```

Comment:
Why the clone? Can you use a `&` instead?

Reply:
It seems like we can use `*` instead.